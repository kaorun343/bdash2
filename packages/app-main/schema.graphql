type BdashQuery {
  body: String!
  createdAt: DateTime!
  id: ID!
  status: QueryStatus
  title: String!
  updatedAt: DateTime!
}

type ConnectionTest {
  message: String
  success: Boolean!
}

input CreateDataSourceSQLite3Input {
  name: String!
  path: String
}

type DataSource {
  config: DataSourceConfig!
  createdAt: DateTime!
  id: ID!
  name: String!
  updatedAt: DateTime!
}

union DataSourceConfig = DataSourceConfigSQLite3

type DataSourceConfigSQLite3 {
  path: String!
}

scalar DateTime

input DeleteDataSourceInput {
  id: ID!
}

type Mutation {
  createDataSourceSQLite3(input: CreateDataSourceSQLite3Input!): DataSource!
  deleteDataSource(input: DeleteDataSourceInput!): Int!
  updateDataSourceSQLite3(input: UpdateDataSourceSQLite3Input!): DataSource!
}

type Query {
  bdashQuery(id: ID!): BdashQuery!
  connectionTestSQLite3(path: String!): ConnectionTest!
  dataSources: [DataSource!]!
  queriesByQueryGroupId(queryGroupId: ID!): [BdashQuery!]!
  queryGroups: [QueryGroup!]!
}

type QueryGroup {
  createdAt: DateTime!
  id: ID!
  name: String!
  updatedAt: DateTime!
}

enum QueryStatus {
  FAILURE
  SUCCESS
}

input UpdateDataSourceSQLite3Input {
  id: ID!
  name: String!
  path: String
}