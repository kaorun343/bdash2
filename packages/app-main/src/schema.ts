import { builder } from './builder'
import './schema/objects/Mutation'
import './schema/objects/Query'

export const schema = builder.toSchema()
