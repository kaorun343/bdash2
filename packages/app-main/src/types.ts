import { ColumnType, Generated, Selectable } from 'kysely'

export type Database = {
  dataSource: DataSourceTable
}

export type DataSourceTable = {
  id: Generated<number>
  name: string
  type: 'sqlite3'
  config: ColumnType<string, string, string>
  createdAt: ColumnType<Date, string, never>
  updatedAt: ColumnType<Date, string, string>
}

export type DataSource = Selectable<DataSourceTable>
