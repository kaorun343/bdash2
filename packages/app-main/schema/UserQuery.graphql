type UserQuery implements Node {
  id: ID!
  title: String!
  body: String!
  createdAt: DateTime!
  updatedAt: DateTime!
  status: UserQueryStatus
}

extend type Query {
  userQuery(id: ID!): UserQuery!
  userQueries: [UserQuery!]!
  userQueriesByGroup(groupId: ID!): [UserQuery!]!
}

input CreateUserQueryInput {
  title: String!
}

type CreateUserQueryPayload {
  userQuery: UserQuery!
}

input UpdateUserQueryTitleInput {
  id: ID!
  title: String!
}

type UpdateUserQueryTitlePayload {
  userQuery: UserQuery!
}

extend type Mutation {
  createUserQuery(input: CreateUserQueryInput!): CreateUserQueryPayload!

  updateUserQueryTitle(input: UpdateUserQueryTitleInput!): UpdateUserQueryTitlePayload!
}
