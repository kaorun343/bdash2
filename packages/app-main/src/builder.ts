import SchemaBuilder from '@pothos/core'
import { Kysely } from 'kysely'
import { Database } from './types'

export const builder = new SchemaBuilder<{
  Context: {
    db: Kysely<Database>
  }
}>({})
