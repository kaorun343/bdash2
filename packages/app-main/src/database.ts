import {
  CompiledQuery,
  DatabaseConnection,
  Dialect,
  Driver,
  Kysely,
  QueryResult,
  SqliteAdapter,
  SqliteIntrospector,
  SqliteQueryCompiler,
} from 'kysely'
import { Database as SqlJsDatabase } from 'sql.js'
import { Database } from './types'

class SqlJsConnection implements DatabaseConnection {
  #db: SqlJsDatabase

  constructor(db: SqlJsDatabase) {
    this.#db = db
  }

  async executeQuery<R>(compiledQuery: CompiledQuery): Promise<QueryResult<R>> {
    const { sql, parameters } = compiledQuery
    const stmt = this.#db.prepare(sql)
    stmt.bind(parameters as any[])

    const rows: R[] = []
    while (stmt.step()) rows.push(stmt.getAsObject() as R)

    return {
      rows,
    }
  }

  async *streamQuery() {
    throw new Error("Sqlite driver doesn't support streaming")
  }
}

export const createDatabase = (database: SqlJsDatabase) => {
  const driver: Driver = {
    init: async () => {},
    acquireConnection: async () => new SqlJsConnection(database),
    beginTransaction: async () => {
      throw new Error('beginTransaction not implemented')
    },
    commitTransaction: async () => {
      throw new Error('commitTransaction not implemented')
    },
    rollbackTransaction: async () => {
      throw new Error('rollbackTransaction not implemented')
    },
    releaseConnection: async () => {},
    destroy: async () => {
      throw new Error('destroy not implemented')
    },
  }

  const dialect: Dialect = {
    createDriver: () => driver,
    createQueryCompiler: () => new SqliteQueryCompiler(),
    createAdapter: () => new SqliteAdapter(),
    createIntrospector: (db: Kysely<Database>) => new SqliteIntrospector(db),
  }

  return new Kysely<Database>({
    dialect,
  })
}
