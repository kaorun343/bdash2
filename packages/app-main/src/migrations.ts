import { Kysely, Migration, Migrator, sql } from 'kysely'

const migrations: Record<string, Migration> = {
  v1: {
    up: async (db: Kysely<any>) => {
      await db.schema
        .createTable('dataSource')
        .addColumn('id', 'integer', (col) => col.autoIncrement().primaryKey())
        .addColumn('name', 'text', (col) => col.notNull())
        .addColumn('type', 'text', (col) => col.notNull())
        .addColumn('config', 'text', (col) => col.notNull())
        .addColumn('createdAt', 'datetime', (col) => col.notNull())
        .addColumn('updatedAt', 'datetime', (col) => col.notNull())
        .execute()

      await db.schema
        .createTable('queryGroup')
        .addColumn('id', 'integer', (col) => col.autoIncrement().primaryKey())
        .addColumn('name', 'text', (col) => col.notNull())
        .addColumn('createdAt', 'datetime', (col) => col.notNull())
        .addColumn('updatedAt', 'datetime', (col) => col.notNull())
        .execute()

      await db.schema
        .createTable('query')
        .addColumn('id', 'integer', (col) => col.autoIncrement().primaryKey())
        .addColumn('queryGroupId', 'integer', (col) => col.references('queryGroup.id').onDelete('set null'))
        .addColumn('dataSourceId', 'integer', (col) => col.references('dataSource.id').onDelete('set null'))
        .addColumn('title', 'text', (col) => col.notNull())
        .addColumn('body', 'text', (col) => col.notNull().defaultTo(''))
        .addColumn('status', 'text', (col) => col.check(sql`status IN ('success', 'failure')`))
        .addColumn('createdAt', 'datetime', (col) => col.notNull())
        .addColumn('updatedAt', 'datetime', (col) => col.notNull())
        .execute()
    },
  },
}

export const migrateToLatest = (db: Kysely<any>) => {
  const migrator = new Migrator({
    db,
    provider: { getMigrations: () => Promise.resolve(migrations) },
  })

  return migrator.migrateToLatest()
}
