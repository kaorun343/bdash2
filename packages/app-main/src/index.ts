import { app, BrowserWindow, ipcMain } from 'electron'
import { graphql } from 'graphql'
import initSqlJs from 'sql.js'
import { createDatabase } from './database'
import { migrateToLatest } from './migrations'
import { schema } from './schema'

const createWindow = (filePath: string) => {
  const win = new BrowserWindow({
    width: 800,
    height: 600,
    webPreferences: {
      preload: `${__dirname}/preload.js`,
    },
  })

  win.loadFile(filePath)
}

export const main = async (filePath: string) => {
  const SQL = await initSqlJs()
  const db = createDatabase(new SQL.Database())

  ipcMain.handle('graphql', async (_event, source: string, variableValues: any) => {
    const result = await graphql({
      schema,
      source,
      contextValue: { db },
      variableValues,
    })
    return result
  })

  await Promise.all([app.whenReady(), migrateToLatest(db)])

  createWindow(filePath)
}
