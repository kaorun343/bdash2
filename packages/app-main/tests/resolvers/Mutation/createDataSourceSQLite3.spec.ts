import initSqlJs from 'sql.js'
import { migrateToLatest } from '../../../src/migrations'
import { Kysely } from 'kysely'
import { Database } from '../../../src/types'
import { createDatabase } from '../../../src/database'
import gql from 'graphql-tag'
import { graphql, print } from 'graphql'
import { schema } from '../../../src/schema'

const SQL = await initSqlJs()
let db: Kysely<Database>

const source = print(gql`
  mutation createDataSourceSQLite3($input: CreateDataSourceSQLite3Input!) {
    createDataSourceSQLite3(input: $input) {
      id
      name
      config {
        __typename
        ... on DataSourceConfigSQLite3 {
          path
        }
      }
    }
  }
`)

beforeEach(async () => {
  db = createDatabase(new SQL.Database())
  await migrateToLatest(db)
})

describe('Mutation.createDataSourceSQLite3', () => {
  it('should create SQLite data source', async () => {
    const result = await graphql({
      schema,
      source,
      contextValue: { db },
      variableValues: { input: { name: 'test', path: '/path/to/sqlite3.database' } },
    })
    expect(result).toMatchSnapshot()
  })
})
