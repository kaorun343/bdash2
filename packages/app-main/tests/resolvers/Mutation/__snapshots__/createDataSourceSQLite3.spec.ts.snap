// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Mutation.createDataSourceSQLite3 > should create SQLite data source 1`] = `
{
  "data": {
    "createDataSourceSQLite3": {
      "config": {
        "__typename": "DataSourceConfigSQLite3",
        "path": "/path/to/sqlite3.database",
      },
      "id": "1",
      "name": "test",
    },
  },
}
`;
